### ISSUE TITLE
J4de - `JUSDBank.sol#_calculateLiquidateAmount` user can front-run liquidate 1 token to prevent others from liquidating

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-04-jojo-judging#221]
J4de

medium

# `JUSDBank.sol#_calculateLiquidateAmount` user can front-run liquidate 1 token to prevent others from liquidating

## Summary

`JUSDBank.sol#_calculateLiquidateAmount` user can front-run repay by 1 token to prevent others from liquidating

## Vulnerability Detail

```solidity
File: Impl/JUSDBank.sol
379     function _calculateLiquidateAmount(
380         address liquidated,
381         address collateral,
382         uint256 amount
383     ) internal view returns (DataTypes.LiquidateData memory liquidateData) {
384         DataTypes.UserInfo storage liquidatedInfo = userInfo[liquidated];
385         require(amount != 0, JUSDErrors.LIQUIDATE_AMOUNT_IS_ZERO);
386         require(
387 >>          amount <= liquidatedInfo.depositBalance[collateral],
388             JUSDErrors.LIQUIDATE_AMOUNT_IS_TOO_BIG
389         );
```

If the liquidation amount is greater than the user's collateral amount during liquidation, revert will occur. Users can expoit this to prevent liquidation:

1. Bob has 100 * 1e18 ETH for collateral, and the position is liquidable
2. Alias is going to liquidate all of Bob's ETH, a total of 100 * 1e18
3. Bob discovered Alias' request, and the front-run liquidated 1 ETH
4. Liquidation fails because 100 * 1e18 > (100 * 1e18 - 1)

P.S. This prevents liquidation itself is invalid, because anyone can have unlimited addresses

```solidity
        require(
            liquidator != liquidated,
            JUSDErrors.SELF_LIQUIDATION_NOT_ALLOWED
        );
```

## Impact

Users can expoit it to prevent being liquidated

## Code Snippet

https://github.com/JOJOexchange/JUSDV1/blob/011e10d36257a404c8c1d7d2b8c9f01a2b7a1969/src/Impl/JUSDBank.sol#L387

## Tool used

Manual Review

## Recommendation

It is recommended to deal with the maximum value when it exceeds

```diff
    function _calculateLiquidateAmount(
        address liquidated,
        address collateral,
        uint256 amount
    ) internal view returns (DataTypes.LiquidateData memory liquidateData) {
        DataTypes.UserInfo storage liquidatedInfo = userInfo[liquidated];
        require(amount != 0, JUSDErrors.LIQUIDATE_AMOUNT_IS_ZERO);
-       require(
-           amount <= liquidatedInfo.depositBalance[collateral],
-           JUSDErrors.LIQUIDATE_AMOUNT_IS_TOO_BIG
-       );
+       if (amount > liquidatedInfo.depositBalance[collateral]) {
+           amount = liquidatedInfo.depositBalance[collateral];
+       }
```

**Comment by thangtranth** [comment]
Escalate for 10 USDC

According to the description, JOJO Exchange is only deployed on Arbitrum. And on Arbitrum, the transactions are ordered by the Sequencer in a first come first served basis. Therefore, front-running is not a valid issue.

Please also review all other issues related to front-running. For example:https://github.com/sherlock-audit/2023-04-jojo-judging/issues/141

For more information:

> The Arbitrum Sequencer orders transactions on a first come, first served basis; the Sequencer inserts transactions into a queue based on the order they are received and executes them accordingly.

https://developer.arbitrum.io/learn-more/faq#does-arbitrum-have-a-mempool

**Comment by sherlock-admin** [comment]
> Escalate for 10 USDC
> 
> According to the description, JOJO Exchange is only deployed on Arbitrum. And on Arbitrum, the transactions are ordered by the Sequencer in a first come first served basis. Therefore, front-running is not a valid issue.
> 
> Please also review all other issues related to front-running. For example:https://github.com/sherlock-audit/2023-04-jojo-judging/issues/141
> 
> For more information:
> 
> > The Arbitrum Sequencer orders transactions on a first come, first served basis; the Sequencer inserts transactions into a queue based on the order they are received and executes them accordingly.
> 
> https://developer.arbitrum.io/learn-more/faq#does-arbitrum-have-a-mempool

You've created a valid escalation for 10 USDC!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by arbitrary-cr0** [comment]
Escalate for 10 USDC

This issue is invalid because not only because Arbitrum uses a centralized sequencer where transactions are ordered on a FCFS (first come first serve) basis, but because the mempool that holds all pending transactions is **private**. Meaning it is currently impossible to frontrun transactions on Arbitrum. Therefore it is impossible for a user to determine if a liquidation transaction is in the mempool, and even if a user somehow did know of a pending liquidation transaction in the mempool, their transaction would be unable to frontrun it due to the FCFS transaction ordering.

**Comment by arbitrary-sparrow** [comment]
Escalate for 10 USDC

Reposting @arbitrary-cr0 so it gets picked up by the bot:

This issue is invalid because not only because Arbitrum uses a centralized sequencer where transactions are ordered on a FCFS (first come first serve) basis, but because the mempool that holds all pending transactions is **private**. Meaning it is currently impossible to frontrun transactions on Arbitrum. Therefore it is impossible for a user to determine if a liquidation transaction is in the mempool, and even if a user somehow did know of a pending liquidation transaction in the mempool, their transaction would be unable to frontrun it due to the FCFS transaction ordering.

**Comment by sherlock-admin** [comment]
> Escalate for 10 USDC
> 
> Reposting @arbitrary-cr0 so it gets picked up by the bot:
> 
> This issue is invalid because not only because Arbitrum uses a centralized sequencer where transactions are ordered on a FCFS (first come first serve) basis, but because the mempool that holds all pending transactions is **private**. Meaning it is currently impossible to frontrun transactions on Arbitrum. Therefore it is impossible for a user to determine if a liquidation transaction is in the mempool, and even if a user somehow did know of a pending liquidation transaction in the mempool, their transaction would be unable to frontrun it due to the FCFS transaction ordering.

You've created a valid escalation for 10 USDC!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by bizzyvinci** [comment]
Escalate for 10 USDC 

This is valid cause based on the contest page, `JOJO is especially interested in losses/malfunctioning related to MEV`. Interested enough to implement a fix.

**Comment by sherlock-admin** [comment]
> Escalate for 10 USDC 
> 
> This is valid cause based on the contest page, `JOJO is especially interested in losses/malfunctioning related to MEV`. Interested enough to implement a fix.

You've created a valid escalation for 10 USDC!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by Trumpero** [comment]
I believe this issue should be invalid since front-running is not possible on Arbitrum.

**Comment by hrishibhat** [comment]
Result:
Invalid
Has duplicates
Based on the comments above on not being able to frontrunning on arbitrum considering this issue invalid

**Comment by sherlock-admin** [comment]
Escalations have been resolved successfully!

Escalation status:
- [thangtranth](https://github.com/sherlock-audit/2023-04-jojo-judging/issues/221/#issuecomment-1569254552): accepted
- [ArbitraryExecution](https://github.com/sherlock-audit/2023-04-jojo-judging/issues/221/#issuecomment-1570559464): accepted
- [bizzyvinci](https://github.com/sherlock-audit/2023-04-jojo-judging/issues/221/#issuecomment-1571257190): rejected

**Comment by IAm0x52** [comment]
Same fix (and issue) as #450
